### PR TITLE
Initial TypeScript implementation

### PR DESCRIPTION
**Pull request by junderw**
In order to add static typing while working on BitcoinJS-lib and automatically generate the type files for use in other TypeScript libraries, I have re-written the library using TypeScript.

A few caveats:

1. Pull requests should modify the ts files, compile, and commit all ts, js, and d.ts files.
2. Before reviewing the JS, a reviewer should build locally and see if a diff occurs.
3. To try and prevent diffs from occurring, setting typescript to a specific version, and only bumping up the version with a re-build and review of the newly compiled JS files (in case of any diff)

ToDos before merging:

1. Thorough review of the TS and JS.
2. Discuss which dependencies we should try and move to TS (at least with definitions, but if this TS methodology above is fine, we could mimic it with other dependencies, ie bip32)
3. Update the dependencies to reflect versions with TS in them. (Currently I have a dependency pointing to the bip32 TS pull request branch)

Any and all feedback is appreciated, and I will try to keep things up to date with master functionality as things change.

I currently have the ES6 PRs and the WitnessBlock checking PRs merged into this branch.

Working on this gave me a new found appreciation for TypeScript.

Any comments are appreciated.

Edit: At this point, diffs kind of lose meaning... so a review would probably consist of:

1. checkout branch
2. npm install
3. npm test (this will perform a build in the process)
4. check to make sure the build didn't cause a diff. (it should be deterministic iirc)
5. check all the files. in src, ts_src, and types
6. check `git diff master -- test/` and see that it's only the new block stuff and a few things to accommodate the new TypeScript structure.

### CODE REVIEW

**Comment by junderw** [comment]
At this point, diffs kind of lose meaning... so a review would probably consist of:

1. checkout branch
2. npm install
3. npm test (this will perform a build in the process)
4. check to make sure the build didn't cause a diff. (it should be deterministic iirc)
5. check all the files. in src, ts_src, and types
6. check `git diff master -- test/` and see that it's only the new block stuff and a few things to accommodate the new TypeScript structure.

**Comment by junderw** [comment]
Related issues / PRs that will be completed if this is merged:
#1314 
#1289 
#1236 
#1200 

Feel free to reference other issues that might be interested in this PR.

File: package.json [hunk 1]
```diff
@@ -14,24 +15,34 @@
...
     "type": "git",
     "url": "https://github.com/bitcoinjs/bitcoinjs-lib.git"
   },
   "files": [
-    "src"
+    "src",
+    "types"
   ],
   "dependencies": {
+    "@types/node": "10.12.18",
```
Comment (johnta0) [inline package.json]: I _prefer_ `@types/node` to be included in devDependencies, or you might have the solid reason to add it to dependencies?

Comment (junderw) [reply package.json]: From my understanding by issues like this:
https://stackoverflow.com/a/46011417

1. Applications should set all types to devDependencies.
2. Libraries should set any types that are exposed to the consumer via themselves in dependencies, and devDependencies should only be for types that are solely used in testing.

That being said... I think most apps will have `@types/node`... so maybe it is ok to move to devDependencies.

File: ts_src/address.ts [hunk 2]
```diff
@@ -0,0 +1,102 @@
+import * as Networks from './networks'
```
[the remark this answers is not in the document]
Comment (junderw) [reply ts_src/address.ts]: Thanks. Could you make a pull request with typeScript branch as base branch?

[the remark this answers is not in the document]
Comment (johnta0) [reply ts_src/address.ts]: I did.
https://github.com/bitcoinjs/bitcoinjs-lib/pull/1323

File: package.json (hunk 1 above)
Comment (dcousens) [reply package.json]: > I think most apps will have @types/node... so maybe it is ok to move to devDependencies.

:+1:

**Comment by d-yokoi** [comment]
@junderw 
Hi, I'm considering adding a formatter and a linter with some basic configurations.
What do you think?

**Comment by junderw** [comment]
great! please make a pr to the typescript branch.

**Comment by d-yokoi** [comment]
please review this pr for using prettier, which is one of the most popular formatters.
https://github.com/bitcoinjs/bitcoinjs-lib/pull/1343

i'll work on a linter after this pr is fixed.

**Comment by junderw** [comment]
@d-yokoi Thanks for the PR, this brings the TypeScript formatting more in line with TypeScript norms.

File: package.json [hunk 3]
```diff
@@ -14,24 +15,35 @@
...
     "bech32": "^1.1.2",
-    "bip32": "^1.0.0",
+    "bip32": "git+https://github.com/junderw/bip32.git#typeScript",
```
Comment (d-yokoi) [inline package.json]: Why don't you push your work to original repository and create a pr so anyone can collaborate on it?

Comment (junderw) [reply package.json]: https://github.com/bitcoinjs/bip32/pull/10

Comment (d-yokoi) [reply package.json]: Sorry, I have missed it.

**Comment by junderw** [comment]
1. After a little more review, we'll merge the BIP32 TS branch and push out a new major version (just in case)
2. Update the dependency here to point to that version.
3. Merge here and bump to v5 just in case.

Probably some time in March.

Also, after merging this I will make a new PR with a concrete upgrade guide from v3 since I still see many people creating issues that amount to "read the new examples."

**Comment by junderw** [comment]
Currently, d-yokoi and myself believe this is ready to merge.
I have pinged some of the devs working on type definitions in Definitely-typed to also get their opinion.

"Like to haves" left:
- TypeScript support for more dependencies.
- Rewrite tests in TypeScript as well.
- `ts_src` folder name change...

**Comment by d-yokoi** [comment]
> ts_src folder name change...

how about `src` and `dist`? this is a more popular naming convention.

**Comment by junderw** [comment]
> how about src and dist? this is a more popular naming convention.

dist implies that the files are only for "distribution" and are not "source" (src) files.

Since we are committing JS to git, JS files should always be checked and verified along side TS files.

dist is more common, but that is because most TS projects do not commit (or even look at) the JS code.

This project is not like that... so a new naming convention is needed.......... but I don't like ts_src... haha

**Comment by youssefgh** [comment]
JS files generated with a specific typescript version + configs should be practically the same, maybe same spaces/tabs/EOL diffs but that shouldn't alter the behavior of the Lib whether a developer use the JS files or the TS ones
I believe that including JS files instead of only generating them when needed would make the review process a bit harder than usual without real added benefits
I couldn't find a reason to include JS in the above comments but maybe am missing some points here
Thank you for the big effort you're putting in this amazing Lib

**Comment by junderw** [comment]
> without real added benefits

Verifiability.

The JS files are committed because people should be given the option to not trust TypeScript/Microsoft if they wish.

TypeScript is only here for two reasons:

1. Makes developing the library more secure, adding static type analysis to the runtime type analysis given by typeforce.
2. Auto-generates type definitions.

An application has the luxury of deciding whether it wants to trust MS.

However, this library is used by people who may want to just verify the JS and not use TS at all. Having git diffs helps that verification.

**Comment by d-yokoi** [comment]
It's common to commit only `.ts` files but publish only `.js` files, which is achieved by using `.gitignore` and `.npmignore` together.

By doing so, users don't necessarily trust TypeScript/Microsoft.

**Comment by youssefgh** [comment]
Now i see the point of keeping the JS files
That make sense for Libs that deal with Bitcoin

**Comment by junderw** [comment]
@d-yokoi That is true, but keeping JS files on git makes it easier for anyone forking us to verify their changes to the ts files are all that is added to the js files. (via git diff)

Plus, that would mean everytime I publish I need to manually verify the whole JS folder.

**Comment by d-yokoi** [comment]
I understand your thoughts.

Then, why don't we leave `.gitignore` as it is and exclude `.ts` files and some config files used only for development from the published package?

Making the size of the package smaller helps users.

**Comment by d-yokoi** [comment]
I have finally noticed `.ts` files and other config files are not to be published from the beginning.

So, please ignore my comment above. Everything looks good to me now :)

**Pull request merged**